File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Columns/SubmenuLeftColumn.cs
using ScaleformUI.Menus;
using ScaleformUI.PauseMenu;

namespace ScaleformUI.PauseMenus.Elements
{
    public class SubmenuLeftColumn : PM_Column
    {
        internal LeftItemType currentItemType => ((TabLeftItem)Items[Index]).ItemType;
        public int CurrentSelection { get => Index; set => Index = value; }
        public TabLeftItem CurrentItem => (TabLeftItem)Items[CurrentSelection];

        public SubmenuLeftColumn(PM_COLUMNS position) : base(position)
        {
            VisibleItems = 10;
        }

        public void AddItem(TabLeftItem item)
        {
            Items.Add(item);
            if (Parent != null && Parent.Visible && Parent.Parent != null && Parent.Parent.Visible)
            {

            }
        }

        public override void SetDataSlot(int index)
        {
            if (index >= Items.Count)
                return;
            TabLeftItem item = (TabLeftItem)Items[index];
            BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DATA_SLOT");
            PushScaleformMovieFunctionParameterInt((int)position);
            PushScaleformMovieFunctionParameterInt(index);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterBool(item.Enabled);
            BeginTextCommandScaleformString("CELL_EMAIL_BCON");
            AddTextComponentScaleform(item.Label);
            EndTextCommandScaleformString_2();
            PushScaleformMovieFunctionParameterBool(false);
            PushScaleformMovieFunctionParameterInt(item.MainColor.ArgbValue);
            PushScaleformMovieFunctionParameterInt(item.HighlightColor.ArgbValue);
            BeginTextCommandScaleformString("CELL_EMAIL_BCON");
            AddTextComponentScaleform(item._internalItem.RightLabel);
            EndTextCommandScaleformString_2();
            PushScaleformMovieFunctionParameterInt((int)item._internalItem.LeftBadge);
            PushScaleformMovieMethodParameterString(item._internalItem.customLeftBadge.Key);
            PushScaleformMovieMethodParameterString(item._internalItem.customLeftBadge.Value);
            PushScaleformMovieFunctionParameterInt((int)item._internalItem.RightBadge);
            PushScaleformMovieMethodParameterString(item._internalItem.customRightBadge.Key);
            PushScaleformMovieMethodParameterString(item._internalItem.customRightBadge.Value);
            PushScaleformMovieMethodParameterString(item._internalItem.labelFont.FontName);
            PushScaleformMovieMethodParameterString(item._internalItem.rightLabelFont.FontName);
            EndScaleformMovieMethod();
        }

        public override void UpdateSlot(int index)
        {
            if (index >= Items.Count)
                return;
            TabLeftItem item = (TabLeftItem)Items[index];
            BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "UPDATE_SLOT");
            PushScaleformMovieFunctionParameterInt((int)position);
            PushScaleformMovieFunctionParameterInt(index);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterBool(item.Enabled);
            BeginTextCommandScaleformString("CELL_EMAIL_BCON");
            AddTextComponentScaleform(item.Label);
            EndTextCommandScaleformString_2();
            PushScaleformMovieFunctionParameterBool(false);
            PushScaleformMovieFunctionParameterInt(item.MainColor.ArgbValue);
            PushScaleformMovieFunctionParameterInt(item.HighlightColor.ArgbValue);
            BeginTextCommandScaleformString("CELL_EMAIL_BCON");
            AddTextComponentScaleform(item._internalItem.RightLabel);
            EndTextCommandScaleformString_2();
            PushScaleformMovieFunctionParameterInt((int)item._internalItem.LeftBadge);
            PushScaleformMovieMethodParameterString(item._internalItem.customLeftBadge.Key);
            PushScaleformMovieMethodParameterString(item._internalItem.customLeftBadge.Value);
            PushScaleformMovieFunctionParameterInt((int)item._internalItem.RightBadge);
            PushScaleformMovieMethodParameterString(item._internalItem.customRightBadge.Key);
            PushScaleformMovieMethodParameterString(item._internalItem.customRightBadge.Value);
            PushScaleformMovieMethodParameterString(item._internalItem.labelFont.FontName);
            PushScaleformMovieMethodParameterString(item._internalItem.rightLabelFont.FontName);
            EndScaleformMovieMethod();
        }

        public override void GoUp()
        {
            Items[Index].Selected = false;
            index--;
            if (index < 0)
                index = Items.Count - 1;
            Items[Index].Selected = true;
            Parent.CenterColumn.Items.Clear();
            if (currentItemType != LeftItemType.Empty)
                Parent.CenterColumn.Items.AddRange(((TabLeftItem)Items[Index]).ItemList);
            if (Parent != null && Parent.Visible && Parent.Parent != null && Parent.Parent.Visible)
                Parent.Parent._pause._pause.CallFunction("MENU_STATE", (int)currentItemType);
        }
         
        public override void GoDown()
        {
            Items[Index].Selected = false;
            index++;
            if (index >= Items.Count)
                index = 0;
            Items[Index].Selected = true;
            Parent.CenterColumn.Items.Clear();
            if (currentItemType != LeftItemType.Empty)
                Parent.CenterColumn.Items.AddRange(((TabLeftItem)Items[Index]).ItemList);
            if (Parent != null && Parent.Visible && Parent.Parent != null && Parent.Parent.Visible)
                Parent.Parent._pause._pause.CallFunction("MENU_STATE", (int)currentItemType);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Columns/SubmenuLeftColumn.lua
SubmenuLeftColumn = {}
SubmenuLeftColumn.__index = SubmenuLeftColumn
setmetatable(SubmenuLeftColumn, { __index = PM_Column })
SubmenuLeftColumn.__call = function() return "SubmenuLeftColumn" end

function SubmenuLeftColumn.New(pos)
    local col = PM_Column.New(pos)
    col.VisibleItems = 10
    return setmetatable(col, SubmenuLeftColumn)
end

function SubmenuLeftColumn:currentItemType()
    return self.Items[self:Index()].ItemType
end

function SubmenuLeftColumn:CurrentItem()
    return self.Items[self:Index()]
end

function SubmenuLeftColumn:AddItem(item)
    item.ParentColumn = self
    table.insert(self.Items, item)
end

function SubmenuLeftColumn:SetDataSlot(index)
    if index > #self.Items then return end
    local item = self.Items[index]
    BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DATA_SLOT")
    PushScaleformMovieFunctionParameterInt(self.position)
    PushScaleformMovieFunctionParameterInt(index - 1)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterBool(item:Enabled())
    BeginTextCommandScaleformString("CELL_EMAIL_BCON")
    AddTextComponentScaleform(item:Label())
    EndTextCommandScaleformString_2()
    PushScaleformMovieFunctionParameterBool(false)
    PushScaleformMovieFunctionParameterInt(item.MainColor:ToArgb())
    PushScaleformMovieFunctionParameterInt(item.HighlightColor:ToArgb())
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieMethodParameterString(ScaleformFonts.CHALET_LONDON_NINETEENSIXTY.FontName)
    PushScaleformMovieMethodParameterString(ScaleformFonts.CHALET_LONDON_NINETEENSIXTY.FontName)
    EndScaleformMovieMethod()
end

function SubmenuLeftColumn:UpdateSlot(index)
    if index > #self.Items then return end
    local item = self.Items[index]
    BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "UPDATE_SLOT")
    PushScaleformMovieFunctionParameterInt(self.position)
    PushScaleformMovieFunctionParameterInt(index - 1)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieFunctionParameterBool(item:Enabled())
    BeginTextCommandScaleformString("CELL_EMAIL_BCON")
    AddTextComponentScaleform(item:Label())
    EndTextCommandScaleformString_2()
    PushScaleformMovieFunctionParameterBool(false)
    PushScaleformMovieFunctionParameterInt(item.MainColor:ToArgb())
    PushScaleformMovieFunctionParameterInt(item.HighlightColor:ToArgb())
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieFunctionParameterInt(0)
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieMethodParameterString("")
    PushScaleformMovieMethodParameterString(ScaleformFonts.CHALET_LONDON_NINETEENSIXTY.FontName)
    PushScaleformMovieMethodParameterString(ScaleformFonts.CHALET_LONDON_NINETEENSIXTY.FontName)
    EndScaleformMovieMethod()
end

function SubmenuLeftColumn:GoUp()
    self.Items[self:Index()]:Selected(false)
    self.index = self.index-1
    if self.index < 1 then
        self.index = #self.Items
    end
    self.Items[self:Index()]:Selected(true)
    self.Parent.CenterColumn:Clear()
    if self:currentItemType() ~= LeftItemType.Empty then
        for k,v in ipairs(self.Items[self:Index()].ItemList) do
            self.Parent.CenterColumn:AddItem(v)
        end
    end
    if self.Parent ~= null and self.Parent.Visible and self.Parent.Parent ~= null and self.Parent.Parent:Visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_STATE", currentItemType);
    end
end

function SubmenuLeftColumn:GoDown()
    self.Items[self:Index()]:Selected(false)
    self.index = self.index+1
    if self.index > #self.Items then
        self.index = 1
    end
    self.Items[self:Index()]:Selected(true)
    self.Parent.CenterColumn:Clear()
    if self:currentItemType() ~= LeftItemType.Empty then
        for k,v in ipairs(self.Items[self:Index()].ItemList) do
            self.Parent.CenterColumn:AddItem(v)
        end
    end
    if self.Parent ~= null and self.Parent.Visible and self.Parent.Parent ~= null and self.Parent.Parent:Visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_STATE", currentItemType);
    end
end